docs/admin/etcd.md

```diff
@@ -1,49 +1,163 @@
-[cluster](/docs/admin/high-availability/#clustering-etcd) (multiple machines each running
-etcd) or etcd's data directory should be located on durable storage (e.g., GCE's
-persistent disk). In either case, if high availability is required--as it might
-be in a production cluster--the data directory ought to be [backed up
-periodically](https://coreos.com/etcd/docs/latest/op-guide/recovery.html),
-to reduce downtime in case of corruption.
+### New etcd docker image
+We decided to completely change the way how etcd image looks like (and how it
+works). So far the docker image for etcd in version X contained only etcd and
+etcdctl binaries in version X.
 
-## Default configuration
+Going forward, the docker image for etcd in version X will contain multiple
+versions of etcd. For example 3.0.14 image will contain all: 2.2.1, 2.3.7 and
+3.0.14 binaries of etcd and etcdctl. This will allow running etcd in multiple
+different versions using exactly the same docker image.
 
-The default setup scripts use kubelet's file-based static pods feature to run etcd in a
-[pod](http://releases.k8s.io/{{page.githubbranch}}/cluster/saltbase/salt/etcd/etcd.manifest). This manifest should only
-be run on master VMs. The default location that kubelet scans for manifests is
-`/etc/kubernetes/manifests/`.
+Additionally, the image will contain our custom script for doing migration
+between versions (written by Kubernetes team) and the custom rollback tool
+(provided by CoreOS folks).
 
-## Kubernetes's usage of etcd
+### Migration script
+The migration script that will now be part of etcd docker image is a bash
+script, working as following:
 
-By default, Kubernetes objects are stored under the `/registry` key in etcd.
-This path can be prefixed by using the [kube-apiserver](/docs/admin/kube-apiserver) flag
-`--etcd-prefix="/foo"`.
+1. Detect which version of etcd we were previously running.
+   For that purpose, we have added a dedicated (metadata) file that will be
+   holding that information and will be stored in etcd-data-specific directory
+   (next to real etcd data). If the file doesn’t exist, we default it to 2.2.1
+   version.
+1. If we are in 2.2.1 version and are supposed to upgrade, backup
+   data.
+1. Based on the detected previous etcd version and the desired one (the desired
+   version is communicated via environment variable), do the upgrade steps (if
+   needed).
+   That means, for every minor etcd release between the detected one (excluding)
+   to the desired one (including):
+   1. start etcd in that version
+   1. wait until it is healthy (healthy means that you can write some data to
+      it)
+   1. stop this etcd
+   The important thing here is that those etcd will not list on the default etcd
+   port.  They are hardcoded to listen on ports which the apiserver is not
+   configured to connect to, which means that apiserver won’t be able to connect
+   to those etcd's and no new data will be written in the meantime.
```

Comment:
Although it _does_ remain part of the etcd cluster, because we don't change the peer port, I believe, so new data will be written.  Should we clarify the "no new data" statement?  (And is this actually safe?)